redis/src/push_manager.rs

```diff
@@ -0,0 +1,254 @@
+                if let Some(channel_id) = channel_id {
+                    sbs.retain(|cs| cs.channel_id != channel_id)
+                } else {
+                    sbs.clear();
+                }
+                if sbs.is_empty() {
+                    subscriptions.remove(&channel_name);
+                    unsubscribed_from_all = true;
+                }
+            }
+            subscriptions
+        });
+        unsubscribed_from_all
+    }
+    fn try_send(&self, val: &Value, con_addr: &Arc<String>) -> RedisResult<()> {
+        let channel_name: String;
+        if let Value::Push { kind, data } = val {
+            let all_subscriptions = self.subscriptions.load();
+            if data.is_empty() {
+                return Err(RedisError::from((
+                    ErrorKind::ResponseError,
+                    "At least one data is expected from this push information",
+                )));
+            }
+            match kind {
+                &PushKind::Message | &PushKind::PMessage | &PushKind::SMessage => {
+                    channel_name = FromRedisValue::from_redis_value(&data[0])?
+                }
+                _ => {
+                    return Err(RedisError::from((
+                        ErrorKind::ResponseError,
+                        "PushKind is unknown",
+                    )));
+                }
+            };
+            let subscriptions = all_subscriptions.get(&channel_name);
+            if let Some(subscriptions) = subscriptions {
+                for subscription in subscriptions {
+                    let push_info = PushInfo {
+                        kind: kind.clone(),
+                        data: data.clone(),
+                        con_addr: con_addr.clone(),
+                    };
+                    let is_closed = match &subscription.sender {
+                        #[cfg(feature = "aio")]
+                        PushSender::Tokio(tokio_sender) => tokio_sender.send(push_info).is_err(),
+                        PushSender::Standard(std_sender) => {
+                            let result = std_sender.try_send(push_info);
+                            matches!(result, Err(TrySendError::Disconnected(_)))
+                        }
+                    };
+                    if is_closed {
+                        //TODO check if it causes any problem when guard is hold
+                        self.unsubscribe_from_channel(
+                            channel_name.clone(),
+                            Some(subscription.channel_id),
+                        );
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+/// Manages Push messages for both tokio and std channels
+#[derive(Clone, Default)]
+pub struct PushManager {
+    senders: Arc<ArcSwap<HashMap<PushKind, Vec<PushSender>>>>,
+    subscriptions: Arc<SubscriptionHolder>,
+    psubscriptions: Arc<SubscriptionHolder>,
```

Comment:
I think it's good to separate and in the future re-subscription can be thing which needs this.

Reply:
why would re-subscription need this? and what benefit do you get from the separation?

Reply:
If they are all mixed and it's not possible to distinguish between pattern/shard/normal channel subscriptions then how to re-subscribe to them ?

Reply:
who re-subscribes, and what do you mean by resubscription?

I'm not sure what's your intent here, but it seems like you're replicating the server state of subscription management. If demuxing the subscriptions necessitates us replicating server state on the client side, IMO it's another argument to avoid demuxing in the library, and leave it to the user.

Reply:
In future ConnectionManager should be able to re-subscribe, but we can leave this to the user too.